### ISSUE TITLE
[BUG] System Backup is stuck at Uploading if there are PVs not provisioned by CSI driver

### ISSUE THREAD

**Issue by haoyixin** [longhorn/longhorn#5286]
## Describe the bug (🐛 if you encounter this issue)

After I create a system backup by ui, it's stucking at `Uploading` state. (But volume backup is working correctly.)

## To Reproduce

Just click create button on system backup page, then type a name and click ok.

## Expected behavior

Successfully create a system backup.

## Log or Support bundle

```
time="2023-01-12T21:02:01Z" level=debug msg="Failed to upload /tmp/test.zip to backup target" controller=longhorn-system-backup error="error uploading system backup from /tmp/test.zip to s3://longhorn@us-east-1/backupstore/system-backups/v1.4.0/test: failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.4.0/longhorn [system-backup upload /tmp/test.zip s3://longhorn@us-east-1/backupstore/system-backups/v1.4.0/test --git-commit bb2aeaa6776c8c17c9f86cf61df27084458f7b3f --manager-image harbor.shopeemobile.com/deep-vespa/longhorn-manager:bb2aeaa6 --engine-image longhornio/longhorn-engine:v1.4.0], output , stderr, time=\"2023-01-12T21:02:01Z\" level=fatal msg=\"Failed to run upload system-backup command\" error=\"failed to get /tmp/test.zip checksum: open /tmp/test.zip: no such file or directory\"\n, error exit status 1" node=10-188-17-4.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=test
... REPEAT HUNDREDS OF ABOVE ...
```

[supportbundle_15bd8f1a-34e2-4c5f-9358-4b901c58106d_2023-01-12T21-23-24Z.zip](https://github.com/longhorn/longhorn/files/10406123/supportbundle_15bd8f1a-34e2-4c5f-9358-4b901c58106d_2023-01-12T21-23-24Z.zip)

## Environment

 - Longhorn version: v1.4.0
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): helm
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: 1.20 - 1.21 (patched)
   - Number of management node in the cluster: 10
   - Number of worker node in the cluster: 68
 - Node config
   - OS type and version: ubuntu 20.04
   - CPU per node: 104
   - Memory per node: 1T
   - Disk type(e.g. SSD/NVMe): NVMe
   - Network bandwidth between the nodes: 50GBps
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Baremetal
 - Number of Longhorn volumes in the cluster: 30

## Additional context

I did a minor patch to support 1.20:
https://github.com/haoyixin/longhorn-manager/commit/bb2aeaa6776c8c17c9f86cf61df27084458f7b3f

**Comment by innobead** [comment]
cc @c3y1huang @longhorn/qa

**Comment by c3y1huang** [comment]
Based on the support bundle, all the managers with `error uploading system backup` restarted during the system backup.
```
> find supportbundle_15bd8f1a-34e2-4c5f-9358-4b901c58106d_2023-01-12T21-23-24Z -type f -exec grep -l "error uploading system backup" {} \; 
supportbundle_15bd8f1a-34e2-4c5f-9358-4b901c58106d_2023-01-12T21-23-24Z/logs/longhorn-system/longhorn-manager-d472l/longhorn-manager.log
supportbundle_15bd8f1a-34e2-4c5f-9358-4b901c58106d_2023-01-12T21-23-24Z/logs/longhorn-system/longhorn-manager-kwpqh/longhorn-manager.log
supportbundle_15bd8f1a-34e2-4c5f-9358-4b901c58106d_2023-01-12T21-23-24Z/logs/longhorn-system/longhorn-manager-bwlb5/longhorn-manager.log
supportbundle_15bd8f1a-34e2-4c5f-9358-4b901c58106d_2023-01-12T21-23-24Z/logs/longhorn-system/longhorn-manager-p6rrr/longhorn-manager.log
supportbundle_15bd8f1a-34e2-4c5f-9358-4b901c58106d_2023-01-12T21-23-24Z/logs/longhorn-system/longhorn-manager-9ckng/longhorn-manager.log
supportbundle_15bd8f1a-34e2-4c5f-9358-4b901c58106d_2023-01-12T21-23-24Z/logs/longhorn-system/longhorn-manager-j8nds/longhorn-manager.log

> kubectl -n longhorn-system get pod | grep longhorn-manager | grep ago\)
haoyixin                                            1/1     Running     1 (9h ago)      37h
longhorn-manager-bwlb5                                            1/1     Running     1 (9h ago)      37h
longhorn-manager-d472l                                            1/1     Running     1 (9h ago)      37h
longhorn-manager-j8nds                                            1/1     Running     1 (9h ago)      37h
longhorn-manager-kwpqh                                            1/1     Running     1 (9h ago)      37h
longhorn-manager-p6rrr                                            1/1     Running     1 (37h ago)     37h
```

@haoyixin can you provide the previous logs of the crashed manager pods? ex: `kubectl -n longhorn-system logs longhorn-manager-bwlb5 --previous`

Since you are using a custom-built image. Have you tried with official v1.4.0 to see if the issue is reproducible? I cannot access `harbor.shopeemobile.com/deep-vespa/longhorn-manager:bb2aeaa6` or `haoyixin/longhorn-manager:bb2aeaa` trying to reproduce this.

**Comment by haoyixin** [comment]
@c3y1huang 
logs by `kubectl -n longhorn-system logs longhorn-manager-bwlb5 --previous`:

```
time="2023-01-12T20:43:09Z" level=info msg="Picked up by SystemBackup Controller 10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com" controller=longhorn-system-backup node=10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=ttt
time="2023-01-12T20:43:10Z" level=info msg="Starting system backup" controller=longhorn-system-backup node=10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=ttt
time="2023-01-12T20:43:10Z" level=info msg="Requeue ttt for next phase: Generating" controller=longhorn-system-backup node=10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=ttt
time="2023-01-12T20:43:10Z" level=info msg="Event(v1.ObjectReference{Kind:\"SystemBackup\", Namespace:\"longhorn-system\", Name:\"ttt\", UID:\"2d4b7014-d443-4ca2-b0bd-b59e9cbd4165\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"66784901\", FieldPath:\"\"}): type: 'Normal' reason: 'Start' Starting system backup"
time="2023-01-12T20:43:10Z" level=info msg="Created system backup file: /tmp/ttt.zip" controller=longhorn-system-backup node=10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=ttt
time="2023-01-12T20:43:10Z" level=info msg="Requeue ttt for next phase: Uploading" controller=longhorn-system-backup node=10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=ttt
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x169dcfb]

goroutine 36063 [running]:
github.com/longhorn/longhorn-manager/datastore.(*DataStore).GetAllPersistentVolumesWithLonghornProvisioner(0xc005264080)
        /go/src/github.com/longhorn/longhorn-manager/datastore/uncached.go:297 +0x17b
github.com/longhorn/longhorn-manager/controller.getObjectsAndPrintToYAML({0xc00439fc60, 0x19}, {0x2088d6b, 0x11}, 0x8, 0x5)
        /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:1031 +0xbb
github.com/longhorn/longhorn-manager/controller.(*SystemBackupController).generateSystemBackupYAMLsForKubernetes(0xc0002b3000, {0xc00439fc60, 0x19})
        /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:778 +0x29c
github.com/longhorn/longhorn-manager/controller.(*SystemBackupController).generateSystemBackupYAMLs(0xc0002b3000, {0xc00563d980, 0xe})
        /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:707 +0x442
github.com/longhorn/longhorn-manager/controller.(*SystemBackupController).generateSystemBackup(0xc006191430, 0xc0029748c0, {0xc006191430, 0x8})
        /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:693 +0xf7
github.com/longhorn/longhorn-manager/controller.(*SystemBackupController).GenerateSystemBackup(0xc0002b3000, 0xc0029748c0, {0xc006191420, 0xc}, {0xc006191430, 0x8})
        /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:665 +0x385
created by github.com/longhorn/longhorn-manager/controller.(*SystemBackupController).reconcile
        /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:419 +0xeb9
```

And I've push my customed image to `hellerhao/longhorn-manager:bb2aeaa6`.

I just made a minor patch https://github.com/haoyixin/longhorn-manager/commit/bb2aeaa6776c8c17c9f86cf61df27084458f7b3f to support `v1.20.15-ske.1` in our production env, I think it's may not affect system backup.

**Comment by c3y1huang** [comment]
> panic logs from `kubectl -n longhorn-system logs longhorn-manager-bwlb5 --previous`:
> 
> ```
> time="2023-01-12T20:43:09Z" level=info msg="Picked up by SystemBackup Controller 10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com" controller=longhorn-system-backup node=10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=ttt
> time="2023-01-12T20:43:10Z" level=info msg="Starting system backup" controller=longhorn-system-backup node=10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=ttt
> time="2023-01-12T20:43:10Z" level=info msg="Requeue ttt for next phase: Generating" controller=longhorn-system-backup node=10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=ttt
> time="2023-01-12T20:43:10Z" level=info msg="Event(v1.ObjectReference{Kind:\"SystemBackup\", Namespace:\"longhorn-system\", Name:\"ttt\", UID:\"2d4b7014-d443-4ca2-b0bd-b59e9cbd4165\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"66784901\", FieldPath:\"\"}): type: 'Normal' reason: 'Start' Starting system backup"
> time="2023-01-12T20:43:10Z" level=info msg="Created system backup file: /tmp/ttt.zip" controller=longhorn-system-backup node=10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=ttt
> time="2023-01-12T20:43:10Z" level=info msg="Requeue ttt for next phase: Uploading" controller=longhorn-system-backup node=10-188-132-3.cls-94kdwxfj.ecp.shopeemobile.com systemBackup=ttt
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x169dcfb]
> 
> goroutine 36063 [running]:
> github.com/longhorn/longhorn-manager/datastore.(*DataStore).GetAllPersistentVolumesWithLonghornProvisioner(0xc005264080)
>         /go/src/github.com/longhorn/longhorn-manager/datastore/uncached.go:297 +0x17b
> github.com/longhorn/longhorn-manager/controller.getObjectsAndPrintToYAML({0xc00439fc60, 0x19}, {0x2088d6b, 0x11}, 0x8, 0x5)
>         /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:1031 +0xbb
> github.com/longhorn/longhorn-manager/controller.(*SystemBackupController).generateSystemBackupYAMLsForKubernetes(0xc0002b3000, {0xc00439fc60, 0x19})
>         /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:778 +0x29c
> github.com/longhorn/longhorn-manager/controller.(*SystemBackupController).generateSystemBackupYAMLs(0xc0002b3000, {0xc00563d980, 0xe})
>         /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:707 +0x442
> github.com/longhorn/longhorn-manager/controller.(*SystemBackupController).generateSystemBackup(0xc006191430, 0xc0029748c0, {0xc006191430, 0x8})
>         /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:693 +0xf7
> github.com/longhorn/longhorn-manager/controller.(*SystemBackupController).GenerateSystemBackup(0xc0002b3000, 0xc0029748c0, {0xc006191420, 0xc}, {0xc006191430, 0x8})
>         /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:665 +0x385
> created by github.com/longhorn/longhorn-manager/controller.(*SystemBackupController).reconcile
>         /go/src/github.com/longhorn/longhorn-manager/controller/system_backup_controller.go:419 +0xeb9
> ```

I see, thank you for reporting. This does look like our bug. Will fix this.

**Comment by innobead** [comment]
@c3y1huang what's the root cause of nil panic?

**Comment by c3y1huang** [comment]
> @c3y1huang what's the root cause of nil panic?

In this case, this cluster has some PV using `hostPath`. When using none-CSI for the source,  the PV will not have the CSI in the spec field. Hence manager crashed with NPE [here](https://github.com/longhorn/longhorn-manager/blob/master/datastore/uncached.go#L297).

**Comment by innobead** [comment]
I see. Let's fix in 1.4.1.

@haoyixin Thanks for the reporting 👍 

cc @longhorn/qa

**Comment by c3y1huang** [comment]
Reproduce Steps
1. Create a workload using `local-path` as the storage class.
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: longhorn-volv-pvc
  namespace: default
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: local-path
  resources:
    requests:
      storage: 5Gi
---
apiVersion: v1
kind: Pod
metadata:
  name: volume-test
  namespace: default
spec:
  restartPolicy: Always
  containers:
  - name: volume-test
    image: nginx:stable-alpine
    imagePullPolicy: IfNotPresent
    livenessProbe:
      exec:
        command:
          - ls
          - /data/lost+found
      initialDelaySeconds: 5
      periodSeconds: 5
    volumeMounts:
    - name: volv
      mountPath: /data
    ports:
    - containerPort: 80
  volumes:
  - name: volv
    persistentVolumeClaim:
      claimName: longhorn-volv-pvc
```
1. Create a Longhorn system backup.
1. System backup cannot complete and the state hangs in `Uploading`.

**Comment by chriscchien** [comment]
Verified in longhorn manager master `1a200c` with [step](https://github.com/longhorn/longhorn/issues/5286#issuecomment-1381686972)
Result: **Pass**

I can reproduce the system backup stuck at uploading status in v1.4.0. In master head,  system backup create success when there exist workload using storageclass `local-path` and `longhorn-static`.

**Comment by willzhang** [comment]
same with openebs-hostpath
```
root@ubuntu:~# kubectl get sc
NAME                 PROVISIONER          RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
longhorn (default)   driver.longhorn.io   Delete          Immediate              true                   44h
openebs-hostpath     openebs.io/local     Delete          WaitForFirstConsumer   false                  4d12h
root@ubuntu:~# kubectl get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                    STORAGECLASS       REASON   AGE
pvc-08ed2170-435e-4863-a7a0-06c45285a382   1Gi        RWO            Delete           Bound    harbor/harbor-jobservice                 openebs-hostpath            3d11h
pvc-1d5c3e7e-b740-4cb8-ab22-65c8e1404bb5   5Gi        RWO            Delete           Bound    harbor/data-harbor-trivy-0               openebs-hostpath            3d11h
pvc-44b3cdff-0739-4894-aee7-730893978e2c   10Gi       RWO            Delete           Bound    wordpress/wordpress                      longhorn                    20h
pvc-5f67c82c-514a-449a-8fd0-b2e5b2c96ba0   10Gi       RWO            Delete           Bound    ingress-nginx1/data-gitea-0              openebs-hostpath            2d23h
pvc-76bda665-2822-4847-9608-91dbe9246eb9   8Gi        RWO            Delete           Bound    tomcat/tomcat                            longhorn                    44h
pvc-8a6de71f-37e8-4d20-b647-36990a622403   5Gi        RWO            Delete           Bound    harbor/harbor-chartmuseum                openebs-hostpath            3d11h
pvc-a2c0f687-e830-478b-8f95-1fa3108dd50b   10Gi       RWO            Delete           Bound    ingress-nginx1/data-gitea-postgresql-0   openebs-hostpath            2d23h
pvc-bca47d01-2f7c-4c56-954c-57816a55ea0e   5Gi        RWO            Delete           Bound    harbor/harbor-registry                   openebs-hostpath            3d11h
pvc-bef9776d-9a49-4958-95e0-e5b9f6b3c4eb   1Gi        RWO            Delete           Bound    harbor/data-harbor-redis-0               openebs-hostpath            3d11h
pvc-c08de926-4c41-496d-b273-55ca5e981c33   1Gi        RWO            Delete           Bound    harbor/harbor-jobservice-scandata        openebs-hostpath            3d11h
pvc-cf9b1ab6-03b2-4a4b-a931-36ec10679365   1Gi        RWO            Delete           Bound    harbor/database-data-harbor-database-0   openebs-hostpath            3d11h
pvc-eea26188-456b-499f-98af-320accf3ecf5   20Gi       RWO            Delete           Bound    lens-metrics/data-prometheus-0           openebs-hostpath            2d1h
pvc-f2e73809-72c5-494f-9617-8c4ebab0a851   8Gi        RWO            Retain           Bound    wordpress/data-wordpress-mariadb-0       longhorn-static             20h
root@ubuntu:~# 
```

**Comment by nlindblo** [comment]
How would I deploy this fix on an existing 1.4.0 system ? Or do I need to wait for a 1.4.1 release

**Comment by khushboo-rancher** [comment]
The solution is available on v1.4.x-head branch but it is recommended that you wait for the upcoming release 1.4.1, which is planned for near future (3-4 weeks from now)
